### PR TITLE
Nodes with a tvu and storage port are no longer double counted

### PR DESCRIPTION
**Pull request by mvines**
`solana-gossip spy --num-nodes 2` can pass when there's only one validator in gossip since a validator has a TVU and Storage port, and each port was considered a "node"

### CODE REVIEW

File: core/src/gossip_service.rs [hunk 1]
```diff
@@ -181,71 +181,55 @@ fn spy(
...
             .unwrap()
             .all_tvu_peers()
             .into_iter()
             .collect::<Vec<_>>();
-        archivers = spy_ref.read().unwrap().all_storage_peers();
+        storage_peers = spy_ref.read().unwrap().all_storage_peers();
+
+        let mut nodes: Vec<_> = tvu_peers.iter().chain(storage_peers.iter()).collect();
+        nodes.sort();
+        nodes.dedup();
```
Comment (mvines) [inline core/src/gossip_service.rs]: Problem: `tvu_peers.len() + archivers.len() >= num` (aka, 1 node is not >= 2 nodes)

Comment (sagar-solana) [reply core/src/gossip_service.rs]: Oops hehe 😅

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Comment by mvines** [comment]
clippy gets me every time 🤬

**Comment by sagar-solana** [comment]
Alternatively we could add another filter for Archivers. They don't have tpu ports so we can filter on having a storage port and definitely not having a tpu port.

File: core/src/gossip_service.rs [hunk 2]
```diff
@@ -181,74 +181,53 @@ fn spy(
...
+                if find_node_by_pubkey.is_none() && find_node_by_gossip_addr.is_none() {
                     met_criteria = true;
-                    break;
                 }
             }
+        } else if found_node_by_pubkey || found_node_by_gossip_addr {
+            met_criteria = true;
         }
-        if i % 20 == 0 {
+        if met_criteria || i % 20 == 0 {
```
Comment (mvines) [inline core/src/gossip_service.rs]: I snuck this change in too -- it's confusing sometimes to not see the final view of gossip in the log before the spy node exits

**Comment by mvines** [comment]
> Alternatively we could add another filter for Archivers. They don't have tpu ports so we can filter on having a storage port and definitely not having a tpu port.

Modifying the filters feels like more risk than I want to take here, mebbie later 🙏

**Comment by mvines** [comment]
Removing v0.23 flag, this isn't needed on v0.23 after all -- the real bug was #8239

**Pull request merged**